golang1.20/lib/launcher_test.go

```diff
@@ -0,0 +1,276 @@
+
+	out, err := execute(f, in)
+	assert.NoError(t, err)
+	assert.Equal(t, string(want), string(out))
+}
+
+func TestInvoke(t *testing.T) {
+	inBytes := []byte(`{"foo":"baz"}`)
+	wantValue := []byte(`{"bar":"baz"}`)
+	wantNothing := []byte("{}")
+	wantError := []byte(fmt.Sprintf(`{"error": %q}`, assert.AnError.Error()))
+
+	tests := []struct {
+		name string
+		f    interface{}
+		want []byte
+	}{{
+		name: "backwards compat",
+		f: func(in map[string]interface{}) map[string]interface{} {
+			return map[string]interface{}{"bar": "baz"}
+		},
+		want: wantValue,
+	}, {
+		name: "func (): return nothing",
+		f:    func() {},
+		want: wantNothing,
+	}, {
+		name: "func () error: return nil",
+		f:    func() error { return nil },
+		want: wantNothing,
+	}, {
+		name: "func () error: return error",
+		f:    func() error { return assert.AnError },
+		want: wantError,
+	}, {
+		name: "func (TIn) error: return nil",
+		f:    func(in TestTypeIn) error { return nil },
+		want: wantNothing,
+	}, {
+		name: "func () (TOut, error): returning value",
+		f: func() (TestTypeOut, error) {
+			return TestTypeOut{Bar: "baz"}, nil
+		},
+		want: wantValue,
+	}, {
+		name: "func () (TOut, error): returning error",
+		f: func() (TestTypeOut, error) {
+			return TestTypeOut{}, assert.AnError
+		},
+		want: wantError,
+	}, {
+		name: "func (Tin) (TOut, error): returning value", // not in Lambdas list
+		f: func(in *TestTypeIn) (*TestTypeOut, error) {
+			return &TestTypeOut{Bar: in.Foo}, nil
+		},
+		want: wantValue,
+	}, {
+		name: "func (Tin) (TOut, error): returning both, error takes precedence",
+		f: func(in *TestTypeIn) (*TestTypeOut, error) {
+			return &TestTypeOut{Bar: in.Foo}, assert.AnError
+		},
+		want: wantError,
+	}, {
+		name: "func (context.Context) error: return nil",
+		f:    func(context.Context) error { return nil },
+		want: wantNothing,
+	}, {
+		name: "func (context.Context) error: return error",
+		f:    func(context.Context) error { return assert.AnError },
+		want: wantError,
+	}, {
+		name: "func (context.Context, Tin) error: return nil",
+		f:    func(context.Context, TestTypeIn) error { return nil },
+		want: wantNothing,
+	}, {
+		name: "func (context.Context, Tin) error: return error",
+		f:    func(context.Context, TestTypeIn) error { return assert.AnError },
+		want: wantError,
+	}, {
+		name: "func (context.Context) (Tout, error): return value",
+		f:    func(context.Context) (TestTypeOut, error) { return TestTypeOut{Bar: "baz"}, nil },
+		want: wantValue,
+	}, {
+		name: "func (context.Context) (Tout, error): return error",
+		f:    func(context.Context) (TestTypeOut, error) { return TestTypeOut{Bar: "baz"}, assert.AnError },
+		want: wantError,
+	}, {
+		name: "func (context.Context, Tin) (Tout, error): return value",
+		f:    func(ctx context.Context, in TestTypeIn) (TestTypeOut, error) { return TestTypeOut{Bar: in.Foo}, nil },
+		want: wantValue,
+	}, {
+		name: "func (context.Context, Tin) (Tout, error): return error",
```

Comment:
not sure if this is applicable here or not, but do we need a test for supporting generic parameters?

Reply:
Well we don't support generic parameters. Not sure that'd even be useful but I'm open to ideas.